syntax = "proto3";

package jiaa.tracking;

option java_multiple_files = true;
option java_package = "com.jiaa.common.tracking";

service TrackingService {
  // 음성 스트리밍 -> STT 변환 (Bidirectional Streaming for Highway AI)
  rpc TranscribeAudio(stream AudioRequest) returns (stream AudioResponse);
  
  // 실행 중인 앱 목록 전송 (변경 시에만)
  rpc SendAppList(AppListRequest) returns (AppListResponse);
}

// ==================== Audio ====================

message AudioRequest {
  bytes audio_data = 1;       // VAD로 걸러진 음성 바이너리
  bool is_final = 2;          // True = "문장 끝났음(침묵 1초 감지)"
  int64 timestamp = 3;        // 로깅용 시간
  string media_info_json = 4; // 미디어 정보 JSON
}

message AudioResponse {
  string transcript = 1;      // 변환된 텍스트 (STT 결과 또는 AI 응답 청크)
  bool is_emergency = 2;      // 위급 상황 여부
  string intent = 3;          // 의도 분석 결과 (JSON) - 마지막 청크에만 포함
  
  // [Highway AI] Streaming 전용 필드
  bool is_partial = 4;        // True = 스트리밍 중간 청크 (TTS 재생용 텍스트)
  bool is_complete = 5;       // True = 마지막 청크 (intent JSON 포함)
  string text_chunk = 6;      // AI 응답 텍스트 조각 (TTS용)
  string emotion = 7;         // 감정 태그 (NORMAL, ANGRY, etc.)
}

// ==================== App List ====================

message AppListRequest {
  string apps_json = 1;       // 실행 중인 앱 목록 JSON (["Chrome", "VS Code", ...])
  int64 timestamp = 2;        // 전송 시간
  
  // 클립보드 데이터 (AES_256_GCM + RSA Hybrid Encrypted)
  // Optional: 클립보드 변경 시에만 채워서 보냄
  bytes clipboard_payload = 3; 
  bytes clipboard_key = 4;
  bytes clipboard_iv = 5;
  bytes clipboard_tag = 6;
}

message AppListResponse {
  bool success = 1;           // 수신 성공 여부
  string message = 2;         // 응답 메시지
  string command = 3;         // 명령 (e.g. "KILL")
  string target_app = 4;      // 대상 앱 (e.g. "Minecraft")
}
